server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -498,7 +491,7 @@ public void onResponse(Boolean forceRefresh) {
 
                     @Override
                     public void onFailure(Exception e) {
-                        refreshFailure.set(e);
+                        refreshed.set(false);
```

Comment:
I am not entirely sure this is correct. The flag we are setting ends up in `forcedRefresh` I think. That flag is not an indication of whether the refresh succeeded. Rather it is whether the indexing operation forced a refresh for various reasons.

But in the timeout case, I suspect we could end up in having a response where the search shard is not necessarily refreshed?

Or are we failing the search shard also on a timeout failure? If so, that seems dangerous.

Reply:
>But in the timeout case, I suspect we could end up in having a response where the search shard is not necessarily refreshed?

In the timeout case we [fail the bulk request](https://github.com/elastic/elasticsearch/blob/7f83189cb24bf41ddea13126385c47302ea9f9b3/server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java#L466-L471) too which seems wrong to me? 

> Or are we failing the search shard also on a timeout failure? If so, that seems dangerous. But given that we apply the timeout [here](https://github.com/elastic/elasticsearch/blob/42dbc3d795c94ed3068dfd45d8b92e110c9076cc/server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java#L149) I think that is not the case.

No, we're not failing the shard after a refresh timeout.